velox/functions/prestosql/DecimalFunctions.cpp

```diff
@@ -364,7 +364,7 @@ void registerDecimalMultiply(const std::string& prefix) {
       exec::SignatureVariable(
           S3::name(),
           fmt::format(
-              "{a_scale} + {b_scale}",
+              "min({a_scale} + {b_scale}, min(38, {a_precision} + {b_precision}))",
```

Comment:
Would it be possible to add a test that fails before the fix to protect this from accidentally breaking in the future?

Reply:
The fuzzer breaks due to it tries to create a DecimalType with invalid scale, and fails on these checks [link](https://github.com/facebookincubator/velox/blob/main/velox/type/Type.h#L713-L721). To test the signature resolving, a invalid DecimalType needs to be used, but it will fail during decimal type creation, instead of during the signature resolving. How do you think? Thank you.

Reply:
@rui-mo Rui, looks like something like this can reproduce the issue.

```
TEST_F(DecimalArithmeticTest, multiply) {
  auto x = makeFlatVector<int128_t>({1000, 2000}, DECIMAL(38, 30));
  auto r = evaluate("c0 * c0", makeRowVector({x}));
```

Reply:
@mbasmanova Thank you. I will take a look.

Reply:
@rui-mo Based on a discussion in https://github.com/prestodb/presto/issues/22383 this may not be the right fix.

Reply:
@mbasmanova Thanks for clarifying. This PR has been updated. Since it is an expected limitation, I feel it might be more proper to handle it in the implementation of decimal multiply. That means the expected behavior of such signature is to throw exception. Do you think it makes sense?

Reply:
@rui-mo It seems strange to me to have a signature that suggests that an operation is supported when it is not supported at all.

What happens if we keep signature as is? I assume function resolution code will fail to resolve the function, which seems like what we want, no?

Perhaps, we need to make sure Fuzzer is prepared to handle this case.

Reply:
> if {a_scale} + {b_scale} exceeds precision, the function fails.

With the result scale defined as {a_scale} + {b_scale}, it seems above behavior is implicit in the signature. When it fails, the error message is like below, which is reported from Type.h, but actually this failure is about the implementation of decimal multiply.

> Exceptions.h:69] Line: ../.././velox/type/Type.h:710, Function:DecimalType, Expression: scale <= precision (46 vs. 38) Scale of decimal type must not exceed its precision, Source: RUNTIME, ErrorCode: INVALID_STATE

So I'm thinking if we can give a more clearer definition for the expected behavior in its implementation.

> Perhaps, we need to make sure Fuzzer is prepared to handle this case.

The fuzzer fails during fuzzing the return type. It seems the fuzzer assumes a return type can be generated for each signature. Without a return type, the expression cannot be created. Maybe, we need to skip this function if exception happens. What do you think? Thank you.

https://github.com/facebookincubator/velox/blob/f0f598642ad9a3e7060377958c67a843b02eacb8/velox/expression/tests/ExpressionFuzzer.cpp#L594

Reply:
> With the result scale defined as {a_scale} + {b_scale}, it seems above behavior is implicit in the signature.

@rui-mo This is indeed the case. There is currently no good way to clearly specify that `{a_scale} + {b_scale}` must be <= 38 and the error message is not very clear. Still, failing at query compilation time might be better then allowing the query to compile, start executing, then fail.

> It seems the fuzzer assumes a return type can be generated for each signature. 

It does looks like this assumption currently exists in the Fuzzer. Maybe we can remove that assumption. If expression doesn't compile, skip it and try a different one.

Reply:
Understood. Thanks for your pointer.

Reply:
Changed this PR to add a test for this case.